### PR TITLE
ROC-1778: Enter defendant type

### PR DESCRIPTION
**Pull request by dharmendrak**

### CODE REVIEW

File: src/main/features/claim/views/defendant-type.njk [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+        {% set errorType = form.errorFor('type') %}
+        <fieldset class="form-group {% if errorType %} form-group-error {% endif %}">
+          <legend class="visuallyhidden">
+            {{ t("Choose defendant type") }}
+          </legend>
+
+          {% if errorType %}<span class="error-message">{{ t(errorType) }}</span>{% endif %}
+          <div class="multiple-choice" data-target="type-ORGANISATION">
+            <input id="ORGANISATION" type="radio" name="type" value="ORGANISATION" {% if form.valueFor('type.value')  == 'ORGANISATION'%} checked="checked" {% endif %}>
```
Comment (robertparkinson) [inline src/main/features/claim/views/defendant-type.njk]: id's shouldn't be upper case. the value can be though

File: src/test/app/forms/models/defendantDetails.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+        expect(errors.length).to.equal(0)
+      })
+    })
+
+    it('should reject defendant details for individual type when individual details are undefined', () => {
+      DefendantTypes.all().forEach(type => {
+        const errors = validator.validateSync(new DefendantDetails(DefendantTypes.INDIVIDUAL, undefined, undefined, undefined, undefined))
+
+        expect(errors.length).to.equal(2)
```
Comment (robertparkinson) [inline src/test/app/forms/models/defendantDetails.ts]: shouldn't this be one error if the title is optional?

File: src/test/app/forms/models/defendantDetails.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        expectValidationError(errors, ValidationErrors.FULLNAME_REQUIRED)
+      })
+    })
+
+    it('should accept defendant details for individual type when individual details are null', () => {
```
Comment (robertparkinson) [inline src/test/app/forms/models/defendantDetails.ts]: this should say reject not accept

File: src/test/app/forms/models/defendantDetails.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+      DefendantTypes.all().forEach(type => {
+        const errors = validator.validateSync(new DefendantDetails(DefendantTypes.INDIVIDUAL, null, null, undefined, undefined))
+
+        expect(errors.length).to.equal(1)
+        expectValidationError(errors, ValidationErrors.FULLNAME_REQUIRED)
+      })
+    })
+
+    it('should accept defendant details for individual type when individual details are null', () => {
```
Comment (robertparkinson) [inline src/test/app/forms/models/defendantDetails.ts]: should say reject not accept and should say when details are empty strings

File: src/test/app/forms/models/defendantDetails.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+      DefendantTypes.all().forEach(type => {
+        const errors = validator.validateSync(new DefendantDetails(DefendantTypes.INDIVIDUAL,
+          randomstring.generate( 36 ), 'fullname', undefined, undefined))
+
+        expect(errors.length).to.equal(1)
+        expectValidationError(errors, ValidationErrors.CONTENT_TOO_LONG)
+      })
+    })
+
+    it('should accept defendant details for individual type when fullName length is 70', () => {
```
Comment (robertparkinson) [inline src/test/app/forms/models/defendantDetails.ts]: and title length is 35 should be added in here

File: src/main/app/drafts/models/draftClaim.ts [hunk 6]
```diff
@@ -1,16 +1,14 @@
...
 import { HousingDisrepair } from 'app/forms/models/housingDisrepair'
 import { PersonalInjury } from 'app/forms/models/personalInjury'
 import PreferredCourt from 'app/forms/models/preferredCourt'
 import Representative from 'drafts/models/representative'
-import Defendant from 'drafts/models/defendant'
+import Defendant from 'app/drafts/models/defendant'
 
 export default class DraftClaim implements Serializable<DraftClaim> {
   claimant: Claimant = new Claimant()
-  summary: Summary = new Summary()
```
Comment (dharmendrak) [inline src/main/app/drafts/models/draftClaim.ts]: not it is part of Claimant

File: src/test/app/forms/models/defendantDetails.ts (hunk 2 above)
Comment (dharmendrak) [inline src/test/app/forms/models/defendantDetails.ts]: Optional just check for null, so there would be two error as undefined and ''  is considered value. Need to work across model

File: src/main/app/drafts/models/defendant.ts [hunk 7]
```diff
@@ -1,11 +1,14 @@
 import { Address } from 'forms/models/address'
+import { DefendantDetails } from 'app/forms/models/defendantDetails'
 
 export default class Defendant {
   address?: Address = new Address()
+  personalDetails?: DefendantDetails = new DefendantDetails()
```
Comment (sabahirfan) [inline src/main/app/drafts/models/defendant.ts]: Just a thought, personalDetails doesn't look appropriate name here. s it contains company house no and organisation. You may call it `defendantDetails` or think something better.

File: src/main/app/drafts/models/draftClaim.ts (hunk 6 above)
Comment (sabahirfan) [inline src/main/app/drafts/models/draftClaim.ts]: By removing summary from here, will break the router `summarise-the-claim.ts`.

File: src/test/app/forms/models/defendantDetails.ts (hunk 3 above)
Comment (dharmendrak) [inline src/test/app/forms/models/defendantDetails.ts]: done

File: src/test/app/forms/models/defendantDetails.ts (hunk 5 above)
Comment (dharmendrak) [inline src/test/app/forms/models/defendantDetails.ts]: it is another test case

File: src/test/app/forms/models/defendantDetails.ts (hunk 4 above)
Comment (dharmendrak) [inline src/test/app/forms/models/defendantDetails.ts]: reworded

File: src/main/features/claim/views/defendant-type.njk (hunk 1 above)
Comment (dharmendrak) [inline src/main/features/claim/views/defendant-type.njk]: changed

File: src/main/features/claim/views/defendant-type.njk [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+            <label for="ORGANISATION">{{ t('Organisation name') }} </label>
+          </div>
+
+          <div
+            class="form-group panel panel-border-narrow js-hidden " id="type-ORGANISATION" aria-hidden="false">
+            {{ textInput('Organisation name', 'organisation', form) }}
```
Comment (sabahirfan) [inline src/main/features/claim/views/defendant-type.njk]: Add translation please for 'Organisation name'

File: src/main/features/claim/views/defendant-type.njk [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+            {{ textInput('Companies House number (optional)', 'companyHouseNumber', form) }}
```
Comment (sabahirfan) [inline src/main/features/claim/views/defendant-type.njk]: Add translation please

File: src/main/features/claim/views/defendant-type.njk [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+          </div>
+
+          <div class="multiple-choice" data-target="type-INDIVIDUAL">
+            <input id="INDIVIDUAL" type="radio" name="type" value="INDIVIDUAL" {% if form.valueFor('type.value')  == 'INDIVIDUAL'%} checked="checked" {% endif %}>
+            <label for="INDIVIDUAL">{{ t('An individual') }} </label>
+          </div>
+          <div
+            class="form-group panel panel-border-narrow js-hidden" id="type-INDIVIDUAL" aria-hidden="false">
+            {{ textInput('Title (Optional)', 'title', form) }}
```
Comment (sabahirfan) [inline src/main/features/claim/views/defendant-type.njk]: Add translation please

File: src/main/features/claim/views/defendant-type.njk [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            {{ textInput('Full name', 'fullName', form) }}
```
Comment (sabahirfan) [inline src/main/features/claim/views/defendant-type.njk]: Add translation please

File: src/test/app/forms/models/defendantDetails.ts (hunk 5 above)
Comment (robertparkinson) [inline src/test/app/forms/models/defendantDetails.ts]: shouldn't we add a test for it so?

File: src/test/app/forms/models/defendantDetails.ts [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+        expectValidationError(errors, ValidationErrors.CONTENT_TOO_LONG)
+      })
+    })
+
+    it('should reject defendant details for individual type when full name length is more than 70', () => {
+      DefendantTypes.all().forEach(type => {
+        const errors = validator.validateSync(new DefendantDetails(DefendantTypes.INDIVIDUAL,
+          undefined, randomstring.generate( 71 ), undefined, undefined))
+
+        expect(errors.length).to.equal(2)
```
Comment (sabahirfan) [inline src/test/app/forms/models/defendantDetails.ts]: shouldn't we be checking the one error here. And if this test case generates two errors, description should mention that.

File: src/test/app/forms/models/defendantDetails.ts [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+      expect(defendantDetails.type).to.equal(undefined)
+      expect(defendantDetails.title).to.equal(undefined)
+      expect(defendantDetails.fullName).to.equal(undefined)
+      expect(defendantDetails.organisation).to.equal(undefined)
+      expect(defendantDetails.companyHouseNumber).to.equal(undefined)
+    })
+
+    it('should have defendant details elements undefined when input has undefined element value', () => {
+      const defendantDetails = DefendantDetails.fromObject({
+        type: null,
```
Comment (sabahirfan) [inline src/test/app/forms/models/defendantDetails.ts]: shouldn't this be undefined?

Comment (dharmendrak) [inline src/test/app/forms/models/defendantDetails.ts]: changed to undefined

File: src/test/app/forms/models/defendantDetails.ts (hunk 12 above)
Comment (dharmendrak) [inline src/test/app/forms/models/defendantDetails.ts]: yes, changed to check one

File: src/main/app/drafts/models/defendant.ts (hunk 7 above)
Comment (dharmendrak) [inline src/main/app/drafts/models/defendant.ts]: renamed to defendantDetails

File: src/main/features/claim/views/defendant-type.njk [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+          {% if errorType %}<span class="error-message">{{ t(errorType) }}</span>{% endif %}
+          <div class="multiple-choice" data-target="type-organisation">
+            <input id="organisationType" type="radio" name="type" value="ORGANISATION" {% if form.valueFor('type.value')  == 'ORGANISATION'%} checked="checked" {% endif %}>
+            <label for="organisationType">{{ t('Organisation name') }} </label>
+          </div>
+
+          <div
+            class="form-group panel panel-border-narrow js-hidden " id="type-organisation" aria-hidden="false">
+            {{ textInput('Organisation name', 'organisation', form) }}
```
Comment (sabahirfan) [inline src/main/features/claim/views/defendant-type.njk]: Translation please

File: src/main/features/claim/views/defendant-type.njk [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+            {{ textInput('Companies House number (optional)', 'companyHouseNumber', form) }}
+          </div>
+
+          <div class="multiple-choice" data-target="type-individual">
+            <input id="individualType" type="radio" name="type" value="INDIVIDUAL" {% if form.valueFor('type.value')  == 'INDIVIDUAL'%} checked="checked" {% endif %}>
+            <label for="individualType">{{ t('An individual') }} </label>
+          </div>
+          <div
+            class="form-group panel panel-border-narrow js-hidden" id="type-individual" aria-hidden="false">
+            {{ textInput('Title (Optional)', 'title', form) }}
```
Comment (sabahirfan) [inline src/main/features/claim/views/defendant-type.njk]: Translation please

File: src/main/features/claim/views/defendant-type.njk [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+            {{ textInput('Full name', 'fullName', form) }}
```
Comment (sabahirfan) [inline src/main/features/claim/views/defendant-type.njk]: Translation please

File: src/test/app/forms/models/defendantDetails.ts (hunk 5 above)
Comment (dharmendrak) [inline src/test/app/forms/models/defendantDetails.ts]: added test for it

File: src/test/app/forms/models/defendantDetails.ts [hunk 17]
```diff
@@ -119,6 +119,15 @@ describe('Defendant Details', () => {
       })
     })
 
+    it('should reject defendant details for organisation type when organisation name length is 255', () => {
```
Comment (sabahirfan) [inline src/test/app/forms/models/defendantDetails.ts]: I think you mean accept here

File: src/test/app/forms/models/defendantDetails.ts [hunk 18]
```diff
@@ -129,6 +138,15 @@ describe('Defendant Details', () => {
       })
     })
 
+    it('should reject defendant details for organisation type when companyHouseNumber length is 8', () => {
```
Comment (sabahirfan) [inline src/test/app/forms/models/defendantDetails.ts]: I think you mean accept here

**Pull request merged**